plugin/src/main/scala/chisel3/internal/plugin/BundleComponent.scala

```diff
@@ -63,62 +97,177 @@ private[plugin] class BundleComponent(val global: Global, arguments: ChiselPlugi
-          val msg = "Users cannot override cloneType. Let the compiler plugin generate it."
+          val msg = "Users cannot override cloneType.  Let the compiler plugin generate it."
           global.globalError(d.pos, msg)
         case _ =>
       }
       (primaryConstructor, paramAccessors.toList)
     }
 
-
     override def transform(tree: Tree): Tree = tree match {
 
-      case bundle: ClassDef if isBundle(bundle.symbol) && !bundle.mods.hasFlag(Flag.ABSTRACT) =>
+      case bundle: ClassDef if isBundle(bundle.symbol) =>
+        def show(string: => String): Unit = {
+          if (
+            !arguments.pluginDebugBundlePattern.isEmpty &&
+            bundle.symbol.name.toString.matches(arguments.pluginDebugBundlePattern)
+          ) {
+            println(string)
+          }
+        }
 
         // ==================== Generate _cloneTypeImpl ====================
         val (con, params) = getConstructorAndParams(bundle.impl.body)
         if (con.isEmpty) {
           global.reporter.warning(bundle.pos, "Unable to determine primary constructor!")
           return super.transform(tree)
         }
-        val constructor = con.get
 
+        val constructor = con.get
         val thiz = gen.mkAttributedThis(bundle.symbol)
 
         // The params have spaces after them (Scalac implementation detail)
         val paramLookup: String => Symbol = params.map(sym => sym.name.toString.trim -> sym).toMap
 
-        // Create a this.<ref> for each field matching order of constructor arguments
-        // List of Lists because we can have multiple parameter lists
-        val conArgs: List[List[Tree]] =
-          constructor.vparamss.map(_.map { vp =>
-            val p = paramLookup(vp.name.toString)
-            // Make this.<ref>
-            val select = gen.mkAttributedSelect(thiz, p)
-            // Clone any Data parameters to avoid field aliasing, need full clone to include direction
-            if (isData(vp.symbol)) cloneTypeFull(select) else select
-          })
-
-        val tparamList = bundle.tparams.map{ t => Ident(t.symbol) }
-        val ttpe = if(tparamList.nonEmpty) AppliedTypeTree(Ident(bundle.symbol), tparamList) else Ident(bundle.symbol)
-        val newUntyped = New(ttpe, conArgs)
-        val neww = localTyper.typed(newUntyped)
-
-        // Create the symbol for the method and have it be associated with the Bundle class
-        val cloneTypeSym =  bundle.symbol.newMethod(TermName("_cloneTypeImpl"), bundle.symbol.pos.focus, Flag.OVERRIDE | Flag.PROTECTED)
-        // Handwritten cloneTypes don't have the Method flag set, unclear if it matters
-        cloneTypeSym.resetFlag(Flags.METHOD)
-        // Need to set the type to chisel3.Bundle for the override to work
-        cloneTypeSym.setInfo(NullaryMethodType(bundleTpe))
-
-        val cloneTypeImpl = localTyper.typed(DefDef(cloneTypeSym, neww))
+        var additionalMethods: Seq[Tree] = Seq()
```

Comment:
Can we do this immutably instead of with a `var`?

Reply:
Done